src/lib/concurrency/logging/group_commit_logger.cpp

```diff
@@ -0,0 +1,275 @@
+ *        all values by iterating over them before putting them into the entry.
+ *        Then the vector needs to be resized just once.
+ *    3.  While writing to disk the buffer is locked with a mutex. A second buffer could be introduced, so log calls can
+ *        be processed in the second buffer while writing the first one to disk.
+ */
+
+#include "group_commit_logger.hpp"
+
+#include <boost/serialization/variant.hpp>
+#include <boost/variant/apply_visitor.hpp>
+
+#include <fcntl.h>
+#include <math.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <chrono>
+#include <fstream>
+#include <future>
+#include <string>
+#include <thread>
+
+#include "binary_recovery.hpp"
+#include "logger.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+// Magic number: buffer size. Buffer is flushed to disk if half of its capacity is exceeded.
+// Therefore this should not be too small, since any log entry needs to fit into half of the buffer.
+constexpr uint32_t LOG_BUFFER_CAPACITY = 16384;
+
+// Magic number: time interval that triggers a flush to disk.
+constexpr auto LOG_INTERVAL = std::chrono::milliseconds(1);
+
+// LogEntry is used to create a single log entry.
+// It keeps track of the cursor position while writing values into a buffer.
+class LogEntry {
+ public:
+  uint32_t cursor{0};
+  std::vector<char> data;
+
+  explicit LogEntry(uint32_t count) { data.resize(count); }
+
+  template <typename T>
+  LogEntry& operator<<(const T& value) {
+    DebugAssert(cursor + sizeof(T) <= data.size(), "Logger: value does not fit into vector, call resize() beforehand");
+    *reinterpret_cast<T*>(&data[cursor]) = value;
+    cursor += sizeof(T);
+    return *this;
+  }
+};
+
+template <>
+LogEntry& LogEntry::operator<<<std::string>(const std::string& value) {
+  DebugAssert(cursor + value.size() < data.size(), "Logger: value does not fit into vector, call resize() beforehand");
+
+  value.copy(&data[cursor], value.size());
+
+  DebugAssert(data[cursor + value.size()] == '\0', "Logger: Byte is not NULL initiated");
+
+  cursor += value.size() + 1;
+
+  return *this;
+}
+
+// EntryWriter is used to write multiple AllTypeVariants into an entry successively.
+// It returns a boolean to indicate if something has been written into entry.
+// Therefore the visitation returns false if the AllTypeVariant is a NullValue.
+// This boolean then is used to set the corresponding bit in the null_value_bitmap.
+// The current implementation resizes the entry for every value.
+// It might improve performance to iterate twice over all values:
+// Acumulate the bytes needed for all values in the first pass,
+// then resize the vector and finally write the values in the second pass.
+class EntryWriter : public boost::static_visitor<bool> {
+ public:
+  explicit EntryWriter(LogEntry& entry) : _entry(entry) {}
+
+  template <typename T>
+  bool operator()(T v) {
+    _entry.data.resize(_entry.data.size() + sizeof(T));
```

Comment:
Running to the allocator every time will kill your performance. Why not reserve something before?